### PR TITLE
tokio-epoll-uring: use it for on-demand downloads

### PR DESCRIPTION
**Pull request by problame**
# Problem

On-demand downloads are still using `tokio::fs`, which we know is inefficient.

# Changes

- Add `pagebench ondemand-download-churn` to quantify on-demand download throughput
  - Requires dumping layer map, which required making `history_buffer` impl `Deserialize`
- Implement an equivalent of `tokio::io::copy_buf` for owned buffers => `owned_buffers_io` module and children.
- Make layer file download sensitive to `io_engine::get()`, using VirtualFile + above copy loop
  - For this, I had to move some code into the `retry_download`, e.g., `sync_all()` call.

Drive-by:
- fix missing escaping in `scripts/ps_ec2_setup_instance_store` 
- if we failed in retry_download to create a file, we'd try to remove it, encounter `NotFound`, and `abort()` the process using `on_fatal_io_error`. This PR adds treats `NotFound` as a success.

# Testing

Functional

- The copy loop is generic & unit tested.

Performance

- Used the `ondemand-download-churn` benchmark to manually test against real S3.
- Results (public Notion page): https://neondatabase.notion.site/Benchmarking-tokio-epoll-uring-on-demand-downloads-2024-04-15-newer-code-03c0fdc475c54492b44d9627b6e4e710?pvs=4
  - Performance is equivalent at low concurrency. Jumpier situation at high concurrency, but, still less CPU / throughput with tokio-epoll-uring.
  - It’s a win.

# Future Work

Turn the manual performance testing described in the above results document into a performance regression test: https://github.com/neondatabase/neon/issues/7146

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -452,6 +452,7 @@ jobs:
       matrix:
         build_type: [ debug, release ]
         pg_version: [ v14, v15, v16 ]
+        pageserver_virtual_file_io_engine: [ std-fs, tokio-epoll-uring ]
```
Comment (problame) [inline .github/workflows/build_and_test.yml:455]: TODO: revert before merge

File: pageserver/src/tenant/remote_timeline_client/download.rs [hunk 2]
```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
...
-            let bytes_amount = tokio::io::copy_buf(&mut reader, &mut destination_file).await;
-
-            match bytes_amount {
-                Ok(bytes_amount) => {
-                    let destination_file = destination_file.into_inner();
-                    Ok((destination_file, bytes_amount))
+            // download to tmp file & make sure tmp file indoe is durable (including file size!)
+            let res = match crate::virtual_file::io_engine::get() {
+                crate::virtual_file::io_engine::IoEngine::NotSet => panic!("unset"),
+                crate::virtual_file::io_engine::IoEngine::StdFs => {
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client/download.rs:82]: Should we move this massive `match` into its own function at least?

Thinking `async fn download_to_tmp_file` or sth

File: pageserver/src/tenant/remote_timeline_client/download.rs [hunk 3]
```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
...
+                Ok(bytes_amount) => Ok(bytes_amount),
                 Err(e) => {
                     if let Err(e) = tokio::fs::remove_file(&temp_file_path).await {
-                        on_fatal_io_error(&e, &format!("Removing temporary file {temp_file_path}"));
+                        if e.kind() != std::io::ErrorKind::NotFound {
+                            on_fatal_io_error(
+                                &e,
+                                &format!("Removing temporary file {temp_file_path}"),
+                            );
+                        }
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client/download.rs:197]: Just highlighting this drive-by-fix, see PR description

> - if we failed in retry_download to create a file, we'd try to remove it, encounter `NotFound`, and `abort()` the process using `on_fatal_io_error`. This PR adds treats `NotFound` as a success.

Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client/download.rs:197]: Thanks for highlighting.

File: pageserver/src/tenant/remote_timeline_client/download.rs [hunk 4]
```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
...
+                        // TODO: use vectored write (writev) once supported by tokio-epoll-uring.
+                        // There's chunks_vectored() on the stream.
+                        let (bytes_amount, destination_file) = async {
+                            let size_tracking = size_tracking_writer::Writer::new(destination_file);
+                            let mut buffered = owned_buffers_io::write::BufferedWriter::<
+                                { super::BUFFER_SIZE },
+                                _,
+                            >::new(size_tracking);
+                            while let Some(res) =
+                                futures::StreamExt::next(&mut download.download_stream).await
```
Comment (koivunej) [inline pageserver/src/tenant/remote_timeline_client/download.rs:155]: How come this does not need pinning?

Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client/download.rs:155]: Aah yes because it's boxpinned already, that's why. I was wondering if this was an older commit from before I lifted the Unpin requirement.

File: pageserver/src/tenant/remote_timeline_client/download.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client/download.rs:82]: perhaps... Thinking of the `download.rs` we have other kinds of download to file which need unifying, like the initdb.tar.zst. So yes, doing this refactoring would help further efforts.

**Review by koivunej — APPROVED** [review]
OwnedAsyncWriter looks good, nice tests around the buffering.

File: pageserver/src/tenant/remote_timeline_client/download.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/tenant/remote_timeline_client/download.rs]: refactor: 54009053a0949e87e242b7f19545afc3bd8373e2

are you suggesting I should switch initdb download as well in this PR?

File: pageserver/src/virtual_file/owned_buffers_io/write.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            if buf.len() >= BUFFER_SIZE {
+                assert_eq!(buf.len(), BUFFER_SIZE);
+                self.flush().await?;
+            }
+        }
+        assert!(chunk.is_empty(), "by now we should have drained the chunk");
+        Ok(())
+    }
+
+    ///
```
Comment (problame) [inline pageserver/src/virtual_file/owned_buffers_io/write.rs:96]: Stray comment

**Comment by problame** [comment]
Only got to do the performance testing now, but, I think this is a fairly safe change, so, merging it now, even if there's less soak time before the next release.
